client/consul/sync.go

```diff
@@ -0,0 +1,341 @@
+func (c *ConsulService) Shutdown() error {
+	var mErr multierror.Error
+	select {
+	case c.shutdownCh <- struct{}{}:
+	default:
+	}
+	for _, service := range c.services {
+		if err := c.client.Agent().ServiceDeregister(service.ID); err != nil {
+			mErr.Errors = append(mErr.Errors, err)
+		}
+	}
+	return mErr.ErrorOrNil()
+}
+
+func (c *ConsulService) RemoveServices(tasks []*structs.Task) error {
+	var mErr multierror.Error
+	var services map[string]struct{}
+	for _, task := range tasks {
+		for _, service := range task.Services {
+			services[service.ID] = struct{}{}
+		}
+	}
+
+	cServices, err := c.client.Agent().Services()
+	if err != nil {
+		return err
+	}
+	cServices = c.filterConsulServices(cServices)
+
+	for _, service := range cServices {
+		if _, validService := services[service.ID]; !validService {
+			if err := c.deregisterService(service.ID); err != nil {
+				mErr.Errors = append(mErr.Errors, err)
+			}
+		}
+	}
+	return mErr.ErrorOrNil()
+}
+
+// registerCheck registers a check definition with Consul
+func (c *ConsulService) registerCheck(check *structs.ServiceCheck, service *consul.AgentService) error {
+	chkReg := consul.AgentCheckRegistration{
+		ID:        check.ID,
+		Name:      check.Name,
+		ServiceID: service.ID,
+	}
+	chkReg.Timeout = check.Timeout.String()
+	chkReg.Interval = check.Interval.String()
+	switch check.Type {
+	case structs.ServiceCheckHTTP:
+		if check.Protocol == "" {
+			check.Protocol = "http"
+		}
+		url := url.URL{
+			Scheme: check.Protocol,
+			Host:   fmt.Sprintf("%s:%d", service.Address, service.Port),
+			Path:   check.Path,
+		}
+		chkReg.HTTP = url.String()
+	case structs.ServiceCheckTCP:
+		chkReg.TCP = fmt.Sprintf("%s:%d", service.Address, service.Port)
+	case structs.ServiceCheckScript:
+		chkReg.TTL = check.Interval.String()
+	}
+	return c.client.Agent().CheckRegister(&chkReg)
+}
+
+// createService creates a Consul AgentService from a Nomad Service
+func (c *ConsulService) createService(service *structs.Service) (*consul.AgentService, error) {
+	host, port := c.task.FindHostAndPortFor(service.PortLabel)
+	if host == "" {
+		return nil, fmt.Errorf("host for the service %q  couldn't be found", service.Name)
+	}
+
+	if port == 0 {
+		return nil, fmt.Errorf("port for the service %q  couldn't be found", service.Name)
+	}
+	srv := consul.AgentService{
+		ID:      service.ID,
+		Service: service.Name,
+		Tags:    service.Tags,
+		Address: host,
+		Port:    port,
+	}
+	return &srv, nil
+}
+
+// registerService registers a service with Consul
+func (c *ConsulService) registerService(service *consul.AgentService) error {
+	srvReg := consul.AgentServiceRegistration{
+		ID:      service.ID,
+		Name:    service.Service,
+		Tags:    service.Tags,
+		Port:    service.Port,
+		Address: service.Address,
+	}
+	return c.client.Agent().ServiceRegister(&srvReg)
+}
+
+// deregisterService de-registers a service with the given ID from consul
+func (c *ConsulService) deregisterService(ID string) error {
```

Comment:
@dadgar helper method so that I don't have to write `c.client.Agent()` everytime.